docs/stateless.md

```diff
@@ -0,0 +1,38 @@
+
+## Overview
+During an IRMA session, the IRMA server acts as a state machine which is described in detail on the [IRMA protocol page](http://localhost:3000/docs/next/irma-protocol#the-session-state). The IRMA session progresses through various states such as `INITIALIZED`, `CONNECTED` and `DONE`. The IRMA server contains logic when to switch from one state into the other. The state itself can be kept in memory or, alternatively, in a Redis data store. Saving the session state in memory requires no extra setup and is currently the default behaviour of the IRMA server. However, the data in memory cannot be shared between several IRMA servers. This means that you cannot scale the IRMA server horizontally when using this in-memory solution. Next to poor scalability, the in-memory solution is problematic when the IRMA server needs to be restarted or - even worse - when the server crashes. In such cases, the in-memory information of all sessions gets irreversibly lost.
+
+With the new feature of running the IRMA server in stateless mode, the session data gets decoupled from the IRMA server, making the server itself stateless. The IRMA server still contains the logic when to switch between states but will retrieve and update the session data from and to a dedicated store. Hence, several IRMA servers can be run in parallel and connect to the same data store. Also, no data is lost when the IRMA server restarts or crashes.
+
+### Use cases
+* Scaling the IRMA server horizontally, i.e. running several IRMA severs in parallel.
+* Preventing data loss when restarting an IRMA server.
+* Preventing data loss in case of IRMA server crashes.
+
+## Running the IRMA server in stateless mode
+### Example
+You can start the IRMA server in stateless mode by setting the `store-type` option to `redis`. Additionally you need to provide a Redis server address and password. For test purposes you can override the need for a password by setting the `redis-allow-empty-password` option to `true`. However, make sure to use a secure Redis password in production. Your Redis data store will contain sensitive data and must be password-protected.
+
+```
+irma server -vv --store-type redis --redis-addr "localhost:6379" --redis-pw "placeholderPassword"
+```
+
+If you want to run several IRMA servers, you can now run them behind a load balancer and connect them to the same Redis instance.
+
+> Currently a simple locking mechanism is implemented. When running several Redis instances in parallel, edge cases may occur. We recommend to use a single Redis instance.
+
+> By default TLS is _not_ enabled when using Redis. Since the session data is not encrypted before being sent to Redis, you should not connect to a Redis instance on an external network.
```

Comment:
This is now more correct, but it seems overly restrictive again; shouldn't it present both options? You either connect to Redis over an internal network, or over an external network using TLS.

Reply:
currently we do not know if we actually support this at all. this is what i meant with the comment that this is still a todo to figure it out.

i can however change the text, so it sounds less restrictive:

```
you should not connect to a Redis instance on an external network.
```

to

```
you should connect to a Redis instance on an internal network.
```

what do you think?

Reply:
I have tried to test it, and found that TLS doesn't work at all with the current implementation: even if I configure Redis with an actual CA-signed TLS certificate, the IRMA server fails to connect to it. It seems that for TLS to be enabled, the [`TLSConfig` field in the `redis.Options`](https://pkg.go.dev/github.com/go-redis/redis/v8@v8.8.0#Options) struct passed to `redis.NewClient()` needs to be set, and we don't currently set that.

So for now, all traffic to Redis will always be in plain text, and then the current negative phrasing (don't do this externally) seems best.